File: src/main/java/org/athenian/battlesnake/snakes/java/SimpleSnake.java
package org.athenian.battlesnake.snakes.java;

import io.battlesnake.core.*;

import java.util.Comparator;
import java.util.List;

import static io.battlesnake.core.JavaConstants.*;
import static org.athenian.battlesnake.snakes.java.SimpleSnake.GameContext;

public class SimpleSnake extends AbstractBattleSnake<GameContext> {

    public static void main(String[] args) {
        new SimpleSnake().run(8080);
    }

    @Override
    public GameContext gameContext() {
        return new GameContext();
    }

    @Override
    public Strategy<GameContext> gameStrategy() {
        return new AbstractStrategy<GameContext>(true) {
            @Override
            public StartResponse onStart(GameContext context, StartRequest request) {
                return new StartResponse("#ff00ff", "beluga", "bolt");
            }

            @Override
            public MoveResponse onMove(GameContext context, MoveRequest request) {
                if (request.isFoodAvailable())
                    return moveTo(request, nearestFood(request.getHeadPosition(), request.getFoodList()));
                else
                    return moveTo(request, request.getBoardCenter());
            }
        };
    }

    MoveResponse moveTo(MoveRequest request, Position position) {
        if (request.getHeadPosition().getX() > position.getX())
            return LEFT;
        else if (request.getHeadPosition().getX() < position.getX())
            return RIGHT;
        else if (request.getHeadPosition().getY() > position.getY())
            return UP;
        else
            return DOWN;
    }

    Position nearestFood(Position headPosition, List<Food> foodList) {
        return foodList.stream()
                .min(Comparator.comparingInt(food -> headPosition.minus(food.getPosition())))
                .get()
                .getPosition();
    }


    static class GameContext extends AbstractGameContext {
    }
}

File: src/main/kotlin/org/athenian/battlesnake/snakes/kotlin/SimpleSnake.kt
package org.athenian.battlesnake.snakes.kotlin

import io.battlesnake.core.*
import org.athenian.battlesnake.snakes.kotlin.SimpleSnake.GameContext

object SimpleSnake : AbstractBattleSnake<GameContext>() {

    class GameContext : AbstractGameContext()

    override fun gameContext() = GameContext()

    override fun gameStrategy() =
        strategy<GameContext>(true) {
            onStart { context: GameContext, request: StartRequest ->
                StartResponse("#ff00ff", "beluga", "bolt")
            }

            onMove { context: GameContext, request: MoveRequest ->
                if (request.isFoodAvailable)
                    moveTo(request, nearestFood(request.headPosition, request.foodList).position)
                else
                    moveTo(request, request.boardCenter)
            }
        }

    fun moveTo(request: MoveRequest, position: Position) =
        when {
            request.headPosition.x > position.x -> LEFT
            request.headPosition.x < position.x -> RIGHT
            request.headPosition.y > position.y -> UP
            else -> DOWN
        }

    fun nearestFood(head: Position, foodList: List<Food>) =
        foodList.maxBy { head - it.position }!!

    @JvmStatic
    fun main(args: Array<String>) {
        run()
    }
}